text/3712-box-yesalias.md

```diff
@@ -0,0 +1,117 @@
+# RFC: No (opsem) Magic Boxes
+
+- Feature Name: `box_yesalias`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#3712](https://github.com/rust-lang/rfcs/pull/3712)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Currently, the operational semantics of the type [`alloc::boxed::Box<T>`](https://doc.rust-lang.org/beta/alloc/boxed/struct.Box.html) is in dispute, but the compiler adds llvm `noalias` to it. To support it, the current operational semantics models have the type use a special form of the `Unique` (Stacked Borrows) or `Active` (Tree Borrows) tag, which has aliasing implications, validity implications, and also presents some unique complications in the model and in improvements to the type (e.g. Custom Allocators). We propose that, for the purposes of the runtime semantics of Rust, `Box` is treated as no more special than a user-defined smart pointer you can write today[^1]. In particular, it is given similar behaviour on a typed copy to a raw pointer.
+
+[^1]: We maintain some trivial validity invariants (such as alignment and address space limits) that a user cannot define, but these invariants only depend upon the value of the `Box` itself, rather than on memory pointed to by the `Box`.
+
+# Motivation
+[motivation]: #motivation
+
+The current behaviour of [`alloc::boxed::Box<T>`] can be suprising, both to unsafe code, and to people working on the language or the compiler. In many respects, `Box<T>` is treated incredibly specially by `rustc` and by Rust, leading to ICEs or unsoundness arising from reasonable changes, such as the introduction of per-container `Allocator`s.
+
+In the past, the operational semantics team has considered many ad-hoc solutions to the problem, while maintaining special cases in the aliasing model (such as Weak Protectors) that only exist for `Box<T>`. 
+For example, moving a `ManuallyDrop<Box<T>>` after calling `Drop` is immediate undefined behaviour (due to the `Box` no longer being dereferenceable) - <https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Moving.20.60ManuallyDrop.3CBox.3C_.3E.3E.60>, and the Active tag requirements for a `Box<T>` are unsound when combined with custom allocators <https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Is.20Box.20a.20language.20type.20or.20a.20library.20type.3F>. This wastes procedural time reviewing the proposals, and complicates the language by introducing special case rules that would otherwise be unnecessary.
```

Comment:
This proposed RFC kind of handwaves over what the opsem complications are that are motivating this.  And for such a consequential RFC, the motivation section is rather short overall.

It'd be helpful for this motivating background context to be inline and explored more deeply here, and for it to go into what the status of the various solutions to these problems might be.

Reply:
E.g., it'd be good to include the substance of @RalfJung's comment [here](https://github.com/rust-lang/rfcs/pull/3712#issuecomment-2421785057) somewhere:

> What happened is that when applying _the Stacked Borrows rules for `Box`_ to `Vec`, a bunch of existing code becomes UB. But that's more a statement about Stacked Borrows than about `noalias`. With Tree Borrows, things are already better, but even Tree Borrows has more UB than LLVM `noalias`. I am not aware of a single real-world usage of `Vec` that actually breaks the `noalias` rules.